### PR TITLE
API Go module

### PR DESCRIPTION
**Pull request by Joerger**
Changes:
- Initialize the `api` module and import it directly into the main `teleport` module.
- Update `make update-vendor` to delete `api` from vendor, and add a symlink in its place.

### CODE REVIEW

**Comment by Joerger** [comment]
@russjones @awly I'm not sure exactly what our vendoring strategy is. How do normally update dependencies, and are the upgrades here an issue?

**Review by a-palchikov — COMMENTED** [review]
Regarding upgrades - these are better done intentionally, e.g. when specific issues are targeted. But regular upgrades are also possible, I guess with the caveat that things need to continue working afterwards.

File: api/go.mod [hunk 1]
```diff
@@ -0,0 +1,19 @@
+module github.com/gravitational/teleport/api
+
+go 1.15
+
+require (
+	github.com/gogo/protobuf v1.3.2
+	github.com/golang/protobuf v1.4.3
+	github.com/gravitational/trace v1.1.14
+	github.com/jonboulle/clockwork v0.2.2 // indirect
+	github.com/sirupsen/logrus v1.7.0 // indirect
```
Comment (russjones) [inline api/go.mod:10]: I know it's indirect, but do you know what's still using `clockwork` and `logrus`?

**Comment by russjones** [comment]
@Joerger I'm not sure if we should vendor `api` here. @awly what do you think?

**Review by awly — COMMENTED** [review]
See https://github.com/golang/go/issues/33789 for the vendoring problem.
Can you try adding some post-processing steps in `make update-vendor` to remove the vendored copy of `api/`?

**Comment by awly** [comment]
See how consul does this: https://github.com/hashicorp/consul/blob/master/GNUmakefile#L225-L228

File: api/go.mod (hunk 1 above)
Comment (Joerger) [reply api/go.mod:10]: Yes, it's from `gravitational/trace`

File: api/go.mod [hunk 2]
```diff
@@ -0,0 +1,19 @@
+module github.com/gravitational/teleport/api
```
[the remark this answers is not in the document]
Comment (Joerger) [reply api/go.mod:1]: I'm going to come back to versioning in a separate PR

**Comment by Joerger** [comment]
> See [golang/go#33789](https://github.com/golang/go/issues/33789) for the vendoring problem.
> Can you try adding some post-processing steps in `make update-vendor` to remove the vendored copy of `api/`?

Post-processing does not seem to be possible in go 1.15, it leads to an "inconsistent vendoring error":

```bash
go: inconsistent vendoring in /home/bjoerger/gravitational/teleport:
	github.com/gravitational/teleport/api@v0.0.0: is explicitly required in go.mod, but not marked as explicit in vendor/modules.txt
	github.com/gravitational/teleport/api: is replaced in go.mod, but not marked as replaced in vendor/modules.txt
``

> See how consul does this: https://github.com/hashicorp/consul/blob/master/GNUmakefile#L225-L228

If you upgrade Consol to 1.14+ you get the same issue:

```bash
go: inconsistent vendoring in /home/bjoerger/examples/consul:
	github.com/hashicorp/consul/api@v1.8.0: is explicitly required in go.mod, but not marked as explicit in vendor/modules.txt
	github.com/hashicorp/consul/sdk@v0.7.0: is explicitly required in go.mod, but not marked as explicit in vendor/modules.txt
	github.com/hashicorp/consul/api: is replaced in go.mod, but not marked as replaced in vendor/modules.txt
	github.com/hashicorp/consul/sdk: is replaced in go.mod, but not marked as replaced in vendor/modules.txt
	github.com/gophercloud/gophercloud@v0.1.0: is marked as explicit in vendor/modules.txt, but not explicitly required in go.mod
```

I spent a while trying to find a way around this in 1.15, but it seems that partial vendoring of modules just isn't supported, and manual modification of `vendor` doesn't seem intended. I couldn't find any other way around vendoring `api`.

**Comment by awly** [comment]
@Joerger I got this working locally with a symlink:
```
.PHONY: update-vendor
update-vendor:
	go mod tidy
	go mod vendor
	rm -r vendor/github.com/gravitational/teleport/api
	ln -s $(shell readlink -f api) vendor/github.com/gravitational/teleport
```

The problem with keeping `api` vendored, is that the compiler looks at the version in `vendor/`. So if you make a change in `api/` and forget to re-vendor, the change won't be picked up.

**Comment by Joerger** [comment]
> @Joerger I got this working locally with a symlink:
> 
> ```
> .PHONY: update-vendor
> update-vendor:
> 	go mod tidy
> 	go mod vendor
> 	rm -r vendor/github.com/gravitational/teleport/api
> 	ln -s $(shell readlink -f api) vendor/github.com/gravitational/teleport
> ```

@awly I'm not sure how to extend this solution to work non-locally, in docker/drone in particular. Is this a good final solution? I can't get anything other than a symlink to have the desired behavior. Another option is to ditch vendoring and depend solely on go mod since that alone should guarentee build consistency, but you may have a better idea as to why we need vendoring. Does teleport need to build without internet?

**Comment by Joerger** [comment]
> @Joerger I got this working locally with a symlink

I changed this to a relative symlink to get it to work in docker without relinking.

**Review by awly — APPROVED** [review]

File: api/types/authentication.go [hunk 3]
```diff
@@ -308,13 +307,13 @@ func (u *U2F) Check() error {
 
 // NewMFADevice creates a new MFADevice with the given name. Caller must set
 // the Device field in the returned MFADevice.
-func NewMFADevice(name string, addedAt time.Time) *MFADevice {
+func NewMFADevice(name, id string, addedAt time.Time) *MFADevice {
```
Comment (Joerger) [inline api/types/authentication.go:310]: @awly please take a quick look at this small refactor to remove the uuid dependency.

File: lib/services/authentication.go [hunk 4]
```diff
@@ -27,6 +27,7 @@ import (
 	"github.com/gravitational/teleport/api/types"
 	"github.com/gravitational/teleport/lib/auth/u2f"
 	"github.com/gravitational/teleport/lib/utils"
+	"github.com/pborman/uuid"
 
 	"github.com/gravitational/trace"
 	"golang.org/x/crypto/bcrypt"
```
Comment (awly) [inline lib/services/authentication.go:33]: nit: group this import below with non-gravitational imports

File: api/types/authentication.go (hunk 3 above)
Comment (awly) [reply api/types/authentication.go:310]: LGTM

**Review by awly — APPROVED** [review]

**Review by r0mant — APPROVED** [review]

**Review by russjones — APPROVED** [review]
Bot.

**Pull request merged**